src/landIce/interfaceWithMPAS/Interface.cpp

```diff
@@ -200,18 +221,70 @@ void velocity_solver_solve_fo(int nLayers, int globalVerticesStride,
-    double* temperature = stk::mesh::field_data(*temperature_field, elem);
-    temperature[0] = temperatureDataOnPrisms[lId];
+  if(depthIntegratedModel) {
+    bool usePOTemp = probParamList.sublist("LandIce Viscosity").get<bool>("Use P0 Temperature");
+    if(!usePOTemp) {
+      // In this case we compute the temperature at quadrature points. 
+      // For each quadrature point we identify what MPAS layer it belongs to, 
+      // and use the temperature at that layer
+      const auto problem = Teuchos::rcp_dynamic_cast<LandIce::StokesFO>(albanyApp->getProblem());
+      TEUCHOS_TEST_FOR_EXCEPTION(Teuchos::is_null(problem), std::runtime_error,
+          "Error! Stokes FO Problem not defined. At the moment the depth integrated model only works for Stokes FO.\n");
+      const auto cellCubature = problem->getCellCubature();
+
+      int numQPs = cellCubature->getNumPoints();
+      Kokkos::DynRankView<double, PHX::Device> quadPointCoords("refPoints", numQPs, 3);
+      Kokkos::DynRankView<double, PHX::Device> quadPointWeights("refWeights", numQPs);
+
+      // Pre-Calculate reference element quantities
+      cellCubature->getCubature(quadPointCoords, quadPointWeights);
+
+      //Compute mesh layers associated to quad points
+      std::vector<int> layerVec(numQPs);
+      for(int qp=0; qp<numQPs; qp++) {
+        int il=0; 
+        auto z = (quadPointCoords(qp,2)+1.0)/2; //quad points are defined on [-1,1]
+        while ((z > levelsNormalizedThickness[il+1]) && (il<nLayers)) il++;
+        layerVec[qp] = il;
+      }    
+      
+      // Populate the temperature mesh field, defined at quad points 
+      QPScalarFieldType* temperature_field = meshStruct->metaData->get_field <QPScalarFieldType> (stk::topology::ELEMENT_RANK, "temperature");
+
+      for(int ib=0; ib <indexToTriangleID.size(); ++ib ) {
+        stk::mesh::Entity elem = meshStruct->bulkData->get_entity(stk::topology::ELEMENT_RANK, indexToTriangleID[ib]);
+        double* temperature = stk::mesh::field_data(*temperature_field, elem);
+        for(int qp=0; qp<numQPs; qp++) {
+          int lId = layerVec[qp] * lElemColumnShift + elemLayerShift * ib;        
+          temperature[qp] = temperatureDataOnPrisms[lId];
+        }
+      }
+    } else {  //P0 temperature
+      ScalarFieldType* temperature_field = meshStruct->metaData->get_field<ScalarFieldType>(stk::topology::ELEMENT_RANK, "temperature");
+      for(int ib=0; ib <indexToTriangleID.size(); ++ib ) {
+        stk::mesh::Entity elem = meshStruct->bulkData->get_entity(stk::topology::ELEMENT_RANK, indexToTriangleID[ib]);
+        double* temperature = stk::mesh::field_data(*temperature_field, elem);
+        for(int il=0; il<nLayers; il++) {
+          int lId = il * lElemColumnShift + elemLayerShift * ib;  
+          double tempFraction = temperatureDataOnPrisms[lId]*(levelsNormalizedThickness[il+1]-levelsNormalizedThickness[il]);
```

Comment:
I'm not surewhat this calculation and the if statment below are about

Reply:
This code computes an average of the MPAS P0 temperature along the column and saves it as a P0 quantity in the Albany one-layer field. I'll add a comment.